### PR TITLE
RavenDB-13212 Concurrency issue when `TryPush` and `TryAcquired` was fixed

### PR DESCRIPTION
**Pull request by haludi**
https://issues.hibernatingrhinos.com/issue/RavenDB-13212

### CODE REVIEW

File: src/Sparrow/Collections/RingBuffer.cs [hunk 1]
```diff
@@ -89,6 +89,13 @@ public bool TryAcquireSingle(out RingItem<T> item)
             }
 
             item = _buffer[sidx & _mask];
+            if (item.Item == null)
```
Comment (ayende) [inline src/Sparrow/Collections/RingBuffer.cs]: Why? That is explicitly something that can happen, this is not an error

File: src/Sparrow/Collections/RingBuffer.cs [hunk 2]
```diff
@@ -50,25 +50,23 @@ public SingleConsumerRingBuffer(int size)
...
                 return false; // No space to do anything
 
-            int ticker = Interlocked.Increment(ref _currentIdx) - 1;
-
             // Assign the value
-            ref var cl = ref _buffer[ticker & _mask];
-
+            ref var cl = ref _buffer[_currentIdx & _mask];
             // We assign the item
             cl.Item = item;
```
Comment (haludi) [inline src/Sparrow/Collections/RingBuffer.cs:70]: The write and the read to the array are not by `Volatile` except for the writing to `RingItem<T>.IsReady.
Is it a problem?

Comment (ayende) [reply src/Sparrow/Collections/RingBuffer.cs:70]: @redknightlois Can you take a look?

**Comment by redknightlois** [comment]
Volatile is needed. This is single consumer / multi producer buffer. The fact that we have a single producer on that case, is not true everywhere

**Comment by redknightlois** [comment]
That would cause on very rare occasions to lose a single push because it will get overwritten during 2 two trypush that happens to read the same idx. If I understand correctly the issue my hypothesis is that we are checking and we allowing to move forward at a risky position while we shouldn't

**Comment by redknightlois** [comment]
After a second look I stand corrected, it is not a rare occurrence, it is very likely on high contention. This is definitely NOT a solution, to solve it look into IsReady interaction with TryAdd

**Comment by ayende** [comment]
@redknightlois any reason not to merge this?

**Comment by redknightlois** [comment]
Yes, it doesnt fix the race condition. It creates a different one far more difficult to diagnose where messages are going to get overwritten (therefore will cause actions to get lost).

**Review by ayende — APPROVED** [review]

**Pull request merged**